[ui] Bugfix the prometheus query to retrieve volume latency

ChengYanJin opened the pull request:
**Component**:  ui, monitoring

**Context**: 
The definition of disk latency is

>   Disk latency is the time that it takes to complete a single I/O operation on a block device


**Summary**:
  const volumeLatencyWriteQuery = 
`sum(irate(node_disk_write_time_seconds_total[5m]) / irate(node_disk_writes_completed_total[5m])) by (instance, device) * 1000000`;

  const volumeLatencyReadQuery = 
`sum(irate(node_disk_read_time_seconds_total[5m]) / irate(node_disk_reads_completed_total[5m])) by (instance, device) * 1000000`;
  
The latency in the volume table:
const volumeLatencyCurrentQuery = 
`irate(node_disk_io_time_seconds_total[1h]) * 1000000`;

 const volumeUsageQuery =
`kubelet_volume_stats_used_bytes / kubelet_volume_stats_capacity_bytes`;

**Acceptance criteria**: 

For loopdevice in Centos 7, the result of Prometheus is always **0**, so `query node_disk_writes_completed_total` and `node_disk_reads_completed_total` should return 0 which cause latency **NaN**, so no graph displayed.

 With warning:
```
WARN Infinite extent for field "read": [Infinity, -Infinity]
WARN Infinite extent for field "write": [Infinity, -Infinity]
```
**Please note that only the last 6 commits are belonging to this PR**

**Please give the approval in PR [#2779](https://github.com/scality/metalk8s/pull/2779)**

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2756

ChengYanJin committed:
ui/tests: Add unit test for addMissingDataPoint functions

Refs:2747

ChengYanJin committed:
ui/tests: Add unit test for getVolumeListData function

Refs: #2747

ChengYanJin committed:
ui/volumes: Remove the utility function jointDataPointBaseonTimeSeries

Since we introduce the aggregate sum function in prometheus query, we don't need to manually merge the values

Refs: #2747

ChengYanJin committed:
ui/volumes: Add props checking for addMissingDataPoint function

Refs: #2747

ChengYanJin committed:
ui/volumes: Aggregate the result of Prometheus queries by `instance` and `device`

- The sum() function provide by Prometheus can group the result
- Use volume capacity return by prometheus to compute the current volume usage
  same for volume usage metrics graph, will address it in the coming PR

Refs: #2747

ChengYanJin committed:
ui/moitoring: Bugfix the latency queries for volumes

The wrong query was `node_disk_io_time_seconds_total`which is the time spend doing I/O (both read and write)

The new latency queries seperate READ/WRITE operations and remove the filter {job="node-exporter"}
- Write latency query: `sum(irate(node_disk_write_time_seconds_total[5m]) / irate(node_disk_writes_completed_total[5m])) by (instance, device)`;
- Read latency query: `sum(irate(node_disk_read_time_seconds_total[5m]) / irate(node_disk_reads_completed_total[5m])) by (instance, device)`;

Explanation of the queries:
- The denominator to compute read latency is `node_disk_reads_completed_total` which is the number of READ operations in this amount of time.
- The numerator to compute read latency is `node_disk_read_time_seconds_total` which is time to do the READ operations

Refs: #2756

ChengYanJin committed:
ui/monitoring: Fix the volume usage query

volume usage query should be 'kubelet_volume_stats_used_bytes / kubelet_volume_stats_capacity_bytes'

Refs: #2756

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

ChengYanJin committed:
ui/tests: Handle the case when the node filter is not valid in the URL

Refs: #2747

ChengYanJin committed:
ui/tests: Handle the case when the `sampleDuration` and `sampleFrequency` is not valid

We have three predefined the durations and frequencies according to the timespan
If the duration and frequency are out of the list, we should return undefined

Refs: #2747

alexandre-allard commented on ui/src/services/NodeVolumesUtils.test.js:
```diff
@@ -420,3 +422,96 @@ it('should return error when called with failed Ready condition', () => {
+      status: 'link',
+      bound: 'Yes',
+      storageCapacity: '5Gi',
+      storageClass: 'metalk8s-prometheus',
```
All occurrences of `metalk8s-prometheus` storage class should probably be replaced by `metalk8s` according to the following PR #2755

ChengYanJin replied:
Got it! Thanks for the sync up!

It will impact the Create volume E2E test!

ChengYanJin committed:
ui/volumes: Return empty array instead of undefined in addMissingDataPoint()

Refs: #2747

ChengYanJin committed:
ui/volumes: Add no data text for the graphs when there is data available

Refs: #2747

ChengYanJin committed:
ui/tests: Update the storage class from `metalk8s-prometheus` to ` metalk8s`

Due to the change of PR #2755, we want to update the storage class

Refs: #2747

slaperche-scality commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -330,22 +335,28 @@ export function* fetchVolumeStats() {
-  const volumeUsedQuery = 'kubelet_volume_stats_used_bytes';
-
-  const volumeUsedQueryResult = yield call(
+  // group the result of the query by instance and device (remove the filter {job="node-exporter"})
```
the `(remove the filter {job="node-exporter"})` should be in a commit message, not in a comment here.

slaperche-scality commented on ui/src/components/VolumeMetricGraphCard.js:
```diff
@@ -161,20 +156,26 @@ const MetricGraphCard = (props) => {
-  const volumeUsageData = volumeUsedOperated?.map((slot) => {
+  const volumeUsageData = volumeUsageOperated?.map((slot) => {
     return {
       date: new Date(slot[0] * 1000), // convert from the RFC 3339 to time in JS
```
Instead of having a comment (which is not repeated l.176 and l.168), we could have a little function `dateFromRFC3339` that you could reuse everywhere it's needed.
And thanks to the explicit name, you won't need the comment.

slaperche-scality commented on ui/src/components/VolumeMetricGraphCard.js:
```diff
@@ -161,20 +156,26 @@ const MetricGraphCard = (props) => {
     return {
       date: new Date(slot[0] * 1000),
-      y: slot[1] === null ? null : Math.round(slot[1] * 1000000),
+      write: slot[1] === null ? null : Math.round(slot[1] * 1000000),
```
ditto, here: instead of having a mysterious `1000000` value used for multiplication, we could hide the computation in a function `toGb` (or a better name :stuck_out_tongue: ) and reuse this function where it's needed.

ChengYanJin replied:
Thanks for the idea! 
The code is more understandable!

slaperche-scality commented on ui/src/services/utils.js:
```diff
@@ -280,7 +280,13 @@ export function addMissingDataPoint(
   sampleFrequency,
 ) {
-  if (!orginalValues || orginalValues.length === 0) {
+  if (
```
could describe which edge cases are handled (in the commit message).

ChengYanJin replied:
hi @slaperche-scality, actually it's more props checking not edge cases, I reword my commit message

slaperche-scality commented on ui/src/containers/VolumePageContent.js:
```diff
@@ -114,30 +114,30 @@ const VolumePageContent = (props) => {
-  );
-  const volumeThroughputWrite = volumeStats?.volumeThroughputWrite?.filter(
+  )?.values;
+  const volumeThroughputWrite = volumeStats?.volumeThroughputWrite?.find(
```
why the change from `filter` to `find`?

ChengYanJin replied:
Since we add the aggregation in the Prometheus query, we have **only one** candidate in the array. 
So `find` is a little bit more efficient than `filter`. Because once it matches the value in findings, it will not check the remaining values in the array collection.

slaperche-scality replied:
Thanks for the explanation :+1: 
This is the kind of detail than can be explained in a commit message :slightly_smiling_face:

slaperche-scality commented on ui/src/components/VolumeMetricGraphCard.js:
```diff
@@ -139,8 +139,11 @@ const MetricGraphCard = (props) => {
   );
-  const volumeLatencyOperated = operateMetricRawData(
-    volumeMetricGraphData?.volumeLatency,
+  const volumeLatencyWriteOperated = operateMetricRawData(
```
Commit message is not very explicit about what was fixed.

You should describe what was the problem with the current code and why the new code address the problem.

ChengYanJin committed:
ui/utils: Add toMega, toMicro and fromUnixTimestampToDate utils functions

Refs: #2756

ChengYanJin commented:
/after_pull_request 2748

slaperche-scality commented on ui/src/components/VolumeMetricGraphCard.js:
```diff
@@ -161,52 +161,59 @@ const MetricGraphCard = (props) => {
-          ? null
-          : Math.round((slot[1] / volumeStorageCapacity) * 100),
+      date: fromUnixTimestampToDate(slot[0]),
+      write: slot[1] === null ? null : Math.round(toMicro(slot[1])),
```
I fact, I think we can ask Prometheus to do the conversion instead of calling `toMicro`.

For instance, using this query:

```javascript
  const volumeLatencyWriteQuery = `sum(irate(node_disk_write_time_seconds_total[5m]) / irate(node_disk_writes_completed_total[5m])) by (instance, device) * 1000000`;
```

To be tested, but I think it works (and if so, same logic can be used for `toMega`).
IMHO, we should let Prometheus do as much work as possible.

gdemonet replied:
> we should let Prometheus do as much work as possible

Should we? Why add load on a critical server-side component when a user's computer is able to do the computation without too much effort? What about the change of scale from the UI (not in scope for now, but wouldn't be absurd)?

slaperche-scality replied:
Because it's Prometheus purpose to deal with those data efficiently (it's a database), so I expect it to be more optimized for this kind of task than some JS code in the browser.
I don't think the multiplication is an heavy operation for Prometheus (otherwise I would be really worried).

If we don't want to load Prometheus, than we should simply fetch raw data and do everything (sampling, aggregation, …) with JS code, no?

> What about the change of scale from the UI (not in scope for now, but wouldn't be absurd)?

The data on the chart are already refreshed regularly, so we would simply update the query we use on the fly.
This won't require extra requests.

gdemonet replied:
Thanks for the explanation, it makes sense indeed. More than a database (some of them are not built with the same purpose in mind), Prometheus is optimized for distributed queries with advanced computations such as the ones you mentioned, so let's use it :+1:!

ChengYanJin committed:
ui: Improvement

Refs: #2747

ChengYanJin committed:
ui/test: Remove the properties `sampleFrequency` and `sampleDuration` limitation in utility function

Improve the test set with human-understandable data instead of the real data

ChengYanJin committed:
ui/tests: Addressed the comments for getVolumeListData function

- Simplfied the state, remove the useless props
- Simplfied the logic in getVolumeListData, since we should have the same behaviour for the wrong Node filter and non-volume Node
- Use camelCase for all the mutable variables

ChengYanJin committed:
ui/monitoring: The job to convert data to micro is done by prometheus

Since prometheus should deal with the data efficiently, we should let it to do as much work as possible
So the convert from second to microsecond should also be done by prometheus

Refs: #2756

ChengYanJin committed:
ui/monitoring: Let Prometheus to convert the data to Mega unit

Refs: #2756

bert-e commented:
# Conflict




There is a conflict between your branch `bugfix/2756-bugfix-the-Prometheus-query-to-retrieve-volume-latency` and the
destination branch `development/2.6`.

Please resolve the conflict on **the feature branch** (`bugfix/2756-bugfix-the-Prometheus-query-to-retrieve-volume-latency`).

```bash
 $ git fetch
 $ git checkout origin/bugfix/2756-bugfix-the-Prometheus-query-to-retrieve-volume-latency
 $ git merge origin/development/2.6
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:bugfix/2756-bugfix-the-Prometheus-query-to-retrieve-volume-latency
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

ChengYanJin committed:
ui/tests: Update the test

Refs: #2756
